org.alloytools.alloy.application/src/main/java/templates/TestTemplateFile.vm

```diff
@@ -412,21 +464,61 @@ class CentralController:
 
     def __init__(self):
         self.concState = $!{rootState.getName()}_State(None, None)
-        self.commands = {"help": self.printHelp, "states": self.printStates, "execute": self.execute}
+        self.concState.default_initialize()
+        self.commands = {"help": self.printHelp, "execute": self.execute}
         #if(${allEnvEvents.size()}> 0)
         self.possibleEvents = {"${allEnvEvents.get(0).getModifiedName()}"#foreach($event in ${allEnvEvents.subList(1, ${allEncEvents.size()})}), "$!{event.getModifiedName()}"#end}
         #else
         self.possibleEvents = {}
         #end
-
+    
+    def handle_event(self):
+        unused_conc_states = self.concState.get_conc_descendants() # set of conc states that haven't taken a transition in this big step
+        conc_state_to_active_child = dict() # dictionary that maps each conc state to its active child at the start of this big step
+        for conc_state in unused_conc_states:
+            conc_state_to_active_child[conc_state] = conc_state.get_active_child_state()
+        prev_num_unused_conc_states = -1
+        
+        while len(unused_conc_states) > 0 and prev_num_unused_conc_states != len(unused_conc_states):
+            prev_num_unused_conc_states = len(unused_conc_states)
+            state_queue = [self.concState]
+            while state_queue:
+                cur_state = state_queue.pop(0)
+
+                possible_transitions = []
+                for k, v in cur_state.transitions.items():
+                    if k():
+                        possible_transitions.append(v)
+
+                # if any transitions are possible, take one at random
+                if possible_transitions:
+                    chosen_transition = random.randint(0, len(possible_transitions) - 1)
+                    possible_transitions[chosen_transition]()
+                    # if conc_state's active state changed (i.e. conc_state took a transition), prevent additional transitions (big step maximality)
```

Comment:
Let's say I have a transition without a goto statement, and if we take this transition, it wouldn't result in a change in active states?